proposals/Reliable Mutability Primitive.md

```diff
@@ -0,0 +1,166 @@
+-->
+People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates. We see new projects are leveraging IPNS over working around it via things like [DNSLink][].
+
+
+##### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+- IPNS has bad reputation, changing it may prove to be a challenge.
+- Teams may find that general primitive is not adequate for custom needs.
+- Required key management (e.g. key recovery, key compromise) still may prove it impractical.
+- Higher level alternatives like textile threads may prove to be better layer of abstraction.
+
+##### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+- Integrating textile threads as core component of the ecosystem.
+- Standardizing DNS record updates mechanism and intergating into IPFS.
+- Work with Pinning Services to implement stardized DNSLink updating mechanism.
+
+##### Dependencies/prerequisites
+<!--List any other projects that are dependencies/prerequisites for this project that is being pitched.-->
+
+- Functioning IPNS in browsers
+- Enabling pubsub in go-ipfs
+
+##### Future opportunities
+<!--What future projects/opportunities could this project enable?-->
+
+- [Petname][] system for memorable but non canonical names (kind of how our phones allow us to not remember numbers).
+- Subscribtion based deny lists (node can subscribe to any endpoint)
+- RSS like systems (In RSS readers memorable namse were not a concern)
+- DAGs revisions carrying IPNS key so that last latest revision or any past revision could be discovered.
+
+### Required resources
+
+
+##### Effort estimate
+<!--T-shirt size rating of the size of the project. If the project might require external collaborators/teams, please note in the roles/skills section below). 
+For a team of 3-5 people with the appropriate skills:
+- Small, 1-2 weeks
+- Medium, 3-5 weeks
+- Large, 6-10 weeks
+- XLarge, >10 weeks
+Describe any choices and uncertainty in this scope estimate. (E.g. Uncertainty in the scope until design work is complete, low uncertainty in execution thereafter.)
+-->
+
+Large. Requires lot of coordinated changes across implementations and collabs. Uncertainty in the scope until concrete actionable plan is in place. Uncertenty in collab buy-in. 
+
+##### Roles / skills needed
+<!--Describe the knowledge/skill-sets and team that are needed for this project (e.g. PM, docs, protocol or library expertise, design expertise, etc.). If this project could be externalized to the community or a team outside PL's direct employment, please note that here.-->
+
+- go-ipfs development
+- js-ipfs development
+- cross org coordinator (to get pinning services onboard)
```

Comment:
Sorry to be slow, but why is it required to get pinning services onboard?  I'm not saying we should skip doing this, I just want to make sure I understand the inter-dependency.  For example, in my head right now I understand that Pinning Services are critical for persistence but I'm less clear on their role in mutability.

Reply:
That is a valid point and calling out pinning services specifically might be projecting too far of how things would work. 

This was partially influenced by the fact that IPNS (at least in its current form) requires records to be republished otherwise they'll become unsolvable. Implicit  assumption here, is that, unless there are motivated nodes to do this, system will end up impractical. Pinning services just seem well positions to do this given that they are motivated nodes keep content that IPNS will point to available (insert git branch analogy here).

That said we could and probably should consider different strategies to meet reliability goals.

Reply:
In other words I think we still need someone to help coordinate efforts beyond just leading implementation efforts, to ensure that we do not end up with a system that is directly tied to PL infrastructure. I am assuming this involves getting other teams onboard.

Reply:
+1 on at least thinking about pinning services in the context of IPNS:

- how pinning of /ipns/{libp2p-key-cid} would work?
- how they can help with avoiding IPNS rot 

etc